org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java

```diff
@@ -308,45 +312,26 @@ protected void incrementalBuild(IResourceDelta delta, final IProgressMonitor mon
 		final ToBeBuilt toBeBuilt = new ToBeBuilt();
 		IResourceDeltaVisitor visitor = createDeltaVisitor(toBeBuiltComputer, toBeBuilt, progress);
 		delta.accept(visitor);
-		processedAbsentReferencedProjects(visitor);
-		if (progress.isCanceled())
+
+		if (progress.isCanceled()) {
 			throw new OperationCanceledException();
-		progress.worked(2);
-		doBuild(toBeBuilt, progress.split(8), BuildType.INCREMENTAL);
-	}
-	
-	/**
-	 * Obtain the deltas for the projects we were previously interested in and process them too, if they do no longer
-	 * have the xtext nature or if they are no longer accessible.
-	 *
-	 * @param visitor
-	 *            the processor.
-	 * @throws CoreException
-	 *             if something goes down the drain.
-	 * @since 2.17
-	 */
-	protected void processedAbsentReferencedProjects(IResourceDeltaVisitor visitor) throws CoreException {
-		final IProject[] interestingProjects = optimisticInvoke("getInterestingProjects");
-		for (IProject more : interestingProjects) {
-			if (!XtextProjectHelper.hasNature(more)) {
-				IResourceDelta interestingDelta = getDelta(more);
-				if (interestingDelta != null) {
-					interestingDelta.accept(visitor);
-				}
-			}
 		}
-	}
-	
-	@SuppressWarnings("unchecked")
-	private <T> T optimisticInvoke(String methodName) {
+		progress.worked(2);
+		Task task = closedProjectsQueue.exhaust();
 		try {
-			return (T) reflector.invoke(this, methodName);
-		} catch (SecurityException | IllegalArgumentException | IllegalAccessException | InvocationTargetException
-				| NoSuchMethodException e) {
-			throw new RuntimeException(e);
+			addInfosFromTask(task, toBeBuilt);
+			doBuild(toBeBuilt, progress.split(8), BuildType.INCREMENTAL);
+		} catch(Exception e) {
+			task.reschedule();
```

Comment:
are we sure this wont produce infinite loops in some circumstances?

Reply:
I am quite sure... but so was I with the previous broken logic ... i will think about a guard, maybe a reschedule counter

Reply:
I am quite sure... but so was I with the previous broken logic ... i will think about a guard, maybe a reschedule counter

Reply:
Done